persistence/src/main/scala/com/mesosphere/usi/storage/zookeeper/PersistenceStore.scala

```diff
@@ -0,0 +1,115 @@
+    */
+  def readFlow: Flow[String, Try[Node], NotUsed]
+  def read(path: String): Future[Try[Node]]
+  def read(paths: Seq[String]): Source[Try[Node], NotUsed] = Source(paths).via(readFlow)
+
+  /**
+    * A Flow for updating nodes in the store. It takes a stream of nodes and returns a stream of paths to indicate
+    * a successful update operation for the returned path.
+    */
+  def updateFlow: Flow[Node, String, NotUsed]
+  def update(node: Node): Future[String]
+  def update(nodes: Seq[Node]): Source[String, NotUsed] = Source(nodes).via(updateFlow)
+
+  /**
+    * A Flow for deleting nodes from the repository. It takes a stream of paths and returns a stream of paths to indicate
+    * a successful deletion operation for the returned path.
+    */
+  def deleteFlow: Flow[String, String, NotUsed]
+  def delete(path: String): Future[String]
+  def delete(paths: Seq[String]): Source[String, NotUsed] = Source(paths).via(deleteFlow)
+
+  /**
+    * Returns the list of paths for children nodes for the passed path. Note that returned path is absolute and contains
+    * node's path prefix.
+    *
+    * @return
+    */
+  type Children = Seq[String]
+  def childrenFlow(absolute: Boolean): Flow[String, Try[Children], NotUsed]
+  def children(path: String, absolute: Boolean): Future[Try[Children]]
+  def children(paths: Seq[String], absolute: Boolean): Source[Try[Children], NotUsed] =
+    Source(paths).via(childrenFlow(absolute))
+
+  /**
+    * Checks for the existence of a node with passed path. The [[existsFlow]] returns a [[scala.Tuple2]] with a path
+    * and a boolean for it's existence.
+    *
+    * @return
+    */
+  def existsFlow: Flow[String, (String, Boolean), NotUsed]
+  def exists(path: String): Future[Boolean]
+  def exists(paths: Seq[String]): Source[(String, Boolean), NotUsed] = Source(paths).via(existsFlow)
+
+  /**
+    * Method syncs state with underlying store.
+    *
+    * @param path node path to sync
+    * @return
+    */
+  def sync(path: String): Future[Done]
+
+  /**
+    * Method takes a list of transaction [[com.mesosphere.usi.storage.zookeeper.PersistenceStore.StoreOp]] operations
+    * and submits them. An exception is thrown if one of the operations fail.
+    */
+  def transaction(operations: Seq[StoreOp]): Future[Done]
+
+  /**
+    * Create a node if none exists with the given path. Return the path of the node. This operation is atomic.
+    *
+    * @param node to create
+    * @return
+    */
+  def createIfAbsent(node: Node): Future[String]
+}
+
+object PersistenceStore {
+
+  /**
+    * Main storage unit: it's a key-value like structure, where the key is the path
+    * to the node e.g. `/app1` which can also be nested e.g. `/apps/app1` and the data
+    * is the value (represented as ByteString)
+    *
+    * @param path node's path
+    * @param data node's payload
+    */
+  case class Node(path: String, data: ByteString)
```

Comment:
I'm not sure this is the interface we want. Checkout `RecordRepository`. The base trait could be templated but I don't know what we gain. We should be fairly specific.

Reply:
This is the **low-level storage interfac**e where (at least for ZK) where we operate with bytes and string paths. 
Not to confuse with the **high-level persistence interface** where we operate with core-types.